content/article/why-i-hate-json-and-you-can-too-.md

```diff
@@ -0,0 +1,177 @@
+
+A secondary displeasure that JSON brings is the annoyance at its rejection
+of trailing commas. While modern JavaScript engines now tolerate these,
+the JSON standard still forbids them, which adds additional overhead when
+maintaining JSON documents.
+
+# 5. Inefficiency: Strings
+
+Every C programmer deals with two fundamental ways of encoding strings:
+either delimit them with a “special” character (in C’s case, the NUL byte)
+or transmit their length along with their “body”.
+
+JSON, being derived from a programming language, uses the former method.
+Unlike C strings, of course, JSON provides an escaping mechanism that allows
+storage of the delimiter character (for JSON, that’s `"`). That, however,
+entails a significant level of complexity for an encoder: a JSON encoder
+cannot encode a string without first inspecting it to look for characters
+that need to be escaped. A string of _n_ bytes can thus be up to _2n_ bytes in
+length, and the only way to know for certain is to inspect each individual
+byte in the string.
+
+It’s far more efficient—particularly for documents that are primarily
+strings—to use a serialization format that encodes string length directly
+and thus obviates any need to escape strings. Both encoding and decoding then
+become a simple copy from one buffer to another.
+
+# 6. Iffy Standardization
+
+I’ve been linking in this article to [json.org](https://json.org) rather than
+a formal
+reference for the format because there **is**, in fact, no unified JSON
+standard. Both the IETF and ECMA maintain JSON standards—the IETF has
+issued no fewer than **three** recensions of theirs!
+
+Even putting that aside, JSON has some peculiar encoding nits. Contrary
+to widespread belief, JSON is _not_ a subset of JavaScript/ECMAScript
+because JSON, unlike JavaScript, permits Unicode line and paragraph
+separators (U+2028 and U+2029, respectively).
+
+JSON is also—again, contrary to widespread misconception—_not_ a subset of
+[YAML](https://yaml.org/), another widely-used serialization format. YAML,
+for example, does not allow the `\/` sequence in encoded strings; this
+sequence is common (though not required) in JSON encoders to avoid HTML
+encoding problems when JSON is output directly into the body of an HTML
+`<script>` tag.
+
+So, JSON is its own little beast.
+
+What to Use Instead?
+=======================
+
+I find [TOML](https://github.com/toml-lang/toml) a compelling format for
+configuration files. It fixes JSON’s UTF-16 problem by allowing both
+`\uXXXX` and `\UXXXXXXXX` formats, and it also allows for comments.
+Projects as diverse as Rust’s [Cargo](https://doc.rust-lang.org/cargo/)
+package manager and [Hugo](https://gohugo.io/)—which powers this site!—are
+some of TOML’s most prominent adopters.
+
+For applications I recommend either the IETF’s [CBOR](https://cbor.io/)
+or Booking.com’s [Sereal](https://github.com/Sereal/Sereal). (Disclaimer:
+I wrote 2 of CPAN’s 3 CBOR libraries.) Both of these
```

Comment:
Here's an idea. Write about CBOR and mention why it's better than JSON :)

But, why is CBOR or Sereal better? Explain the problems they solve. Again, show some examples. Gotta get some Perl in here.

Reply:
wilco